lib/experimental/cmdline.nim

```diff
@@ -0,0 +1,2819 @@
+##   --help           display help message
+##   -f, --force      force removal
+##   -r, --recursive  recursively remove files
+## ```
+
+type
+  Cli*[T] {.requiresInit.} = object
+    ## A command line interface description.
+    ##
+    ## See also:
+    ## - `initCli proc <#initCli,sinkCommandBuilder[T]>`_
+    command: Table[CommandId, CliCommand]
+      ## Lookup mapping of command to lookup tables.
+    parser: Store[ParameterId, ParserAny[T]]
+      ## Parser to process input for ParameterId.
+    name: Store[ParameterId, string]
+      ## Canonical names for all ParameterIds.
+    alias: Table[ParameterId, seq[string]]
+      ## Mapping of ParameterId to aliases.
+    parent: Store[ParameterId, ParameterId]
+      ## Mapping of ParameterId to their parent.
+
+    # Documentation storage
+    #
+    # Might be useful to support not having these for space-constrained
+    # targets using a define.
+    usage: Store[ParameterId, string] ## Canonical usage for ParameterIds.
+    placeholder: Store[ParameterId, string]
+      ## Canonical placeholder for ParameterIds. Only used for flags.
+
+  CliCommand = object
+    ## Lookup table for flags and positionals.
+    flag: Table[string, ParameterId]
+      ## Lookup mapping of flag names to ParameterId.
+    command: Table[string, CommandId]
+      ## Lookup mapping of command names to CommandId.
+    positional: seq[ParameterId]
+      ## Lookup mapping of position to ParameterId.
+
+  FlagBuilder*[T] = object
+    ## Builder for command line flags.
+    ##
+    ## See also:
+    ## - `flagBuilder proc <#flagBuilder,Cli[T]>`_
+    flagParser: ParserAny[T]
+    flagName: string
+    aliases: seq[string]
+    usage: string
+    placeholder: string
+
+  PositionalBuilder*[T] = object
+    ## Builder for command line positional parameters.
+    ##
+    ## See also:
+    ## - `positionalBuilder proc <#positionalBuilder,Cli[T]>`_
+    posParser: ParserAny[T]
+    posName: string
+    usage: string
+
+  CommandBuilder*[T] = object
+    ## Builder for command line subcommands.
+    ##
+    ## See also:
+    ## - `commandBuilder proc <#commandBuilder,Cli[T]>`_
+    cmdParser: ParserAny[T]
+    cmdName: string
+    aliases: seq[string]
+    usage: string
+    isDefault: bool
+
+  Action* {.pure.} = enum
+    ## Action to be taken after parsing.
+    Continue ## Continue parameter parsing.
+    ShowHelp ## Abort and show help message.
+    DisableFlagProcessing ## ParameterIds following this will be considered
+                          ## to be values.
+
+  MaybeAction* = Action | void
+    ## Typeclass to support parsers returning either `Action` or nothing.
+
+  FlagParser*[T; R: MaybeAction] = proc (option, value: string, accumulator: var T): R
+    ## A parser for flag with name `option` and value `value`. The accumulator
+    ## passed to `run()`_ or `parse()`_ can be accessed and modified via
+    ## `accumulator`.
+    ##
+    ## A parser may choose to return an `Action` or not return anything,
```

Comment:
I known that inlining is not possible, but the doc comment duplication is a bit much, and I'd prefer if the expectations of a parser are put under a "Writing Parsers" (or something along that line) section in the top-level documentation, to which every parser type then links.

Autocomplete user won't have immediate access to the information then (unless using an IDE that supports following links in documentation boxes), but I don't think it's super important for the types.